File: packages/sunrey-chain/src/wallet/address.ts
/**
 * Versioned SunRey Blockchain address format.
 *
 * Canonical binary is 42 bytes. Canonical text is HRP + '1' + base32.
 * The address binds network class, algorithm, and a SHA-256 payload of
 * the public-key or account descriptor. Cross-network reuse is rejected.
 *
 * Do not hard-code a single cryptographic algorithm forever.
 */

import { createHash } from 'node:crypto';

import {
  ADDRESS_ALGORITHM_IDS,
  ADDRESS_ALGORITHMS,
  ADDRESS_CLASS_IDS,
  ADDRESS_CLASSES,
  ADDRESS_FORMAT_VERSION,
  ADDRESS_MAX_BINARY_BYTES,
  ADDRESS_MAX_TEXT_LENGTH,
  DEVELOPMENT_NETWORK_IDS,
  NETWORK_CLASS_IDS,
  NETWORK_CLASSES,
  RESERVED_PRODUCTION_NETWORK_ID,
  TESTNET_NETWORK_IDS,
  type AccountDescriptor,
  type AddressAlgorithm,
  type AddressClass,
  type BlockchainAddress,
  type NetworkClass,
  type PublicKeyDescriptor,
} from './types.ts';

const MAGIC = Buffer.from('SR');
const PAYLOAD_DOMAIN = 'SUNREY-ADDR-PAYLOAD-V1';
const CHECKSUM_DOMAIN = 'SUNREY-ADDR-V1';
const BASE32 = 'abcdefghijklmnopqrstuvwxyz234567';

const HRP_BY_CLASS: { readonly [K in NetworkClass]: string } = {
  DEVELOPMENT: 'srdev',
  RESERVED_TEST: 'srtst',
  RESERVED_PRODUCTION: 'srprd',
};

export function networkClassOf(networkId: string): NetworkClass | null {
  if ((DEVELOPMENT_NETWORK_IDS as readonly string[]).includes(networkId)) {
    return 'DEVELOPMENT';
  }
  if ((TESTNET_NETWORK_IDS as readonly string[]).includes(networkId) || networkId.startsWith('net_sunrey_testnet_')) {
    return 'RESERVED_TEST';
  }
  if (
    networkId === RESERVED_PRODUCTION_NETWORK_ID ||
    networkId.startsWith('net_sunrey_production_candidate_')
  ) {
    return 'RESERVED_PRODUCTION';
  }
  return null;
}

export function addressClassFromId(id: number): AddressClass | null {
  const found = (Object.entries(ADDRESS_CLASS_IDS) as Array<[AddressClass, number]>).find(
    ([, value]) => value === id,
  );
  return found ? found[0] : null;
}

export function addressAlgorithmFromId(id: number): AddressAlgorithm | null {
  const found = (Object.entries(ADDRESS_ALGORITHM_IDS) as Array<[AddressAlgorithm, number]>).find(
    ([, value]) => value === id,
  );
  return found ? found[0] : null;
}

export function networkClassFromId(id: number): NetworkClass | null {
  const found = (Object.entries(NETWORK_CLASS_IDS) as Array<[NetworkClass, number]>).find(
    ([, value]) => value === id,
  );
  return found ? found[0] : null;
}

function sha256(...parts: readonly (string | Uint8Array)[]): Buffer {
  const hash = createHash('sha256');
  for (const part of parts) {
    hash.update(part);
  }
  return hash.digest();
}

export function descriptorPayload(
  networkId: string,
  addressClass: AddressClass,
  algorithm: AddressAlgorithm,
  descriptorBytes: Uint8Array,
): Buffer {
  return sha256(
    PAYLOAD_DOMAIN,
    networkId,
    addressClass,
    algorithm,
    descriptorBytes,
  );
}

export function publicKeyDescriptorBytes(descriptor: PublicKeyDescriptor): Buffer {
  return Buffer.from(
    `${descriptor.schemaVersion}|${descriptor.keyId}|${descriptor.suiteId}|${descriptor.algorithm}|${descriptor.publicKeyHex}|${descriptor.purpose}`,
    'utf8',
  );
}

export function accountDescriptorBytes(descriptor: AccountDescriptor): Buffer {
  return Buffer.from(
    `${descriptor.schemaVersion}|${descriptor.accountId}|${descriptor.addressClass}|${descriptor.authorizedKeyIds.join(',')}|${descriptor.policyKind}|${descriptor.threshold}`,
    'utf8',
  );
}

function encodeBase32(bytes: Uint8Array): string {
  let bits = 0;
  let value = 0;
  let out = '';
  for (const byte of bytes) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      out += BASE32[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) {
    out += BASE32[(value << (5 - bits)) & 31];
  }
  return out;
}

function decodeBase32(text: string): Buffer | null {
  let bits = 0;
  let value = 0;
  const out: number[] = [];
  for (const char of text) {
    const idx = BASE32.indexOf(char);
    if (idx < 0) {
      return null;
    }
    value = (value << 5) | idx;
    bits += 5;
    if (bits >= 8) {
      out.push((value >>> (bits - 8)) & 255);
      bits -= 8;
    }
  }
  return Buffer.from(out);
}

function checksumOf(prefix: Uint8Array): Buffer {
  return sha256(CHECKSUM_DOMAIN, prefix).subarray(0, 4);
}

export function encodeAddress(input: {
  readonly networkId: string;
  readonly addressClass: AddressClass;
  readonly algorithm: AddressAlgorithm;
  readonly descriptorBytes: Uint8Array;
}): BlockchainAddress {
  const networkClass = networkClassOf(input.networkId);
  if (!networkClass) {
    throw new TypeError(`unknown network id ${input.networkId}`);
  }
  if (!(ADDRESS_CLASSES as readonly string[]).includes(input.addressClass)) {
    throw new TypeError('unknown address class');
  }
  if (!(ADDRESS_ALGORITHMS as readonly string[]).includes(input.algorithm)) {
    throw new TypeError('unknown address algorithm');
  }
  const payload = descriptorPayload(
    input.networkId,
    input.addressClass,
    input.algorithm,
    input.descriptorBytes,
  );
  const prefix = Buffer.alloc(38);
  MAGIC.copy(prefix, 0);
  prefix[2] = ADDRESS_FORMAT_VERSION;
  prefix[3] = NETWORK_CLASS_IDS[networkClass];
  prefix[4] = ADDRESS_CLASS_IDS[input.addressClass];
  prefix[5] = ADDRESS_ALGORITHM_IDS[input.algorithm];
  payload.copy(prefix, 6);
  const checksum = checksumOf(prefix);
  const binary = Buffer.concat([prefix, checksum]);
  if (binary.length !== ADDRESS_MAX_BINARY_BYTES) {
    throw new TypeError('canonical address binary must be 42 bytes');
  }
  const body = encodeBase32(
    Buffer.concat([
      Buffer.from([ADDRESS_FORMAT_VERSION, prefix[4], prefix[5]]),
      payload,
      checksum,
    ]),
  );
  const text = `${HRP_BY_CLASS[networkClass]}1${body}`;
  if (text.length > ADDRESS_MAX_TEXT_LENGTH) {
    throw new TypeError(`address text exceeds ${ADDRESS_MAX_TEXT_LENGTH} characters`);
  }
  return Object.freeze({
    schemaVersion: ADDRESS_FORMAT_VERSION,
    text,
    binaryHex: binary.toString('hex'),
    networkId: input.networkId,
    networkClass,
    addressClass: input.addressClass,
    algorithm: input.algorithm,
    payloadHex: payload.toString('hex'),
  });
}

export function encodeFromPublicKey(
  networkId: string,
  addressClass: AddressClass,
  descriptor: PublicKeyDescriptor,
): BlockchainAddress {
  return encodeAddress({
    networkId,
    addressClass,
    algorithm: descriptor.algorithm,
    descriptorBytes: publicKeyDescriptorBytes(descriptor),
  });
}

export function encodeFromAccountDescriptor(
  networkId: string,
  algorithm: AddressAlgorithm,
  descriptor: AccountDescriptor,
): BlockchainAddress {
  return encodeAddress({
    networkId,
    addressClass: descriptor.addressClass,
    algorithm,
    descriptorBytes: accountDescriptorBytes(descriptor),
  });
}

export type AddressParseError =
  | 'WRONG_NETWORK_ADDRESS'
  | 'CHECKSUM_FAILURE'
  | 'MALFORMED_ADDRESS'
  | 'UNKNOWN_VERSION'
  | 'UNKNOWN_CLASS'
  | 'UNKNOWN_ALGORITHM';

export type AddressParseResult =
  | { readonly ok: true; readonly address: BlockchainAddress }
  | { readonly ok: false; readonly code: AddressParseError; readonly detail: string };

export function parseAddress(
  text: string,
  expectedNetworkId?: string,
): AddressParseResult {
  const sep = text.indexOf('1');
  if (sep < 4 || text.length > ADDRESS_MAX_TEXT_LENGTH) {
    return { ok: false, code: 'MALFORMED_ADDRESS', detail: 'address text is malformed' };
  }
  const hrp = text.slice(0, sep);
  const body = text.slice(sep + 1);
  const networkClass = (Object.entries(HRP_BY_CLASS) as Array<[NetworkClass, string]>).find(
    ([, value]) => value === hrp,
  )?.[0];
  if (!networkClass || !(NETWORK_CLASSES as readonly string[]).includes(networkClass)) {
    return { ok: false, code: 'WRONG_NETWORK_ADDRESS', detail: 'human-readable prefix is not a SunRey network' };
  }
  const decoded = decodeBase32(body);
  if (!decoded || decoded.length < 3 + 32 + 4) {
    return { ok: false, code: 'MALFORMED_ADDRESS', detail: 'base32 payload is incomplete' };
  }
  const version = decoded[0];
  const classId = decoded[1];
  const algId = decoded[2];
  const payload = decoded.subarray(3, 35);
  const checksum = decoded.subarray(35, 39);
  if (version === undefined || classId === undefined || algId === undefined) {
    return { ok: false, code: 'MALFORMED_ADDRESS', detail: 'base32 payload is incomplete' };
  }
  if (version !== ADDRESS_FORMAT_VERSION) {
    return { ok: false, code: 'UNKNOWN_VERSION', detail: `unsupported address version ${version}` };
  }
  const addressClass = addressClassFromId(classId);
  const algorithm = addressAlgorithmFromId(algId);
  if (!addressClass) {
    return { ok: false, code: 'UNKNOWN_CLASS', detail: 'reserved address class requires a governed upgrade' };
  }
  if (!algorithm) {
    return { ok: false, code: 'UNKNOWN_ALGORITHM', detail: 'unknown address algorithm' };
  }
  const prefix = Buffer.alloc(38);
  MAGIC.copy(prefix, 0);
  prefix[2] = version;
  prefix[3] = NETWORK_CLASS_IDS[networkClass];
  prefix[4] = classId;
  prefix[5] = algId;
  payload.copy(prefix, 6);
  if (!checksum.equals(checksumOf(prefix))) {
    return { ok: false, code: 'CHECKSUM_FAILURE', detail: 'address checksum failed' };
  }
  if (expectedNetworkId) {
    const expectedClass = networkClassOf(expectedNetworkId);
    if (expectedClass !== networkClass) {
      return { ok: false, code: 'WRONG_NETWORK_ADDRESS', detail: 'address network class does not match expected network' };
    }
  }
  const binary = Buffer.concat([prefix, checksum]);
  const inferredNetworkId =
    expectedNetworkId ??
    (networkClass === 'DEVELOPMENT'
      ? DEVELOPMENT_NETWORK_IDS[0]
      : networkClass === 'RESERVED_TEST'
        ? TESTNET_NETWORK_IDS[1]
        : RESERVED_PRODUCTION_NETWORK_ID);
  return {
    ok: true,
    address: Object.freeze({
      schemaVersion: ADDRESS_FORMAT_VERSION,
      text,
      binaryHex: binary.toString('hex'),
      networkId: inferredNetworkId,
      networkClass,
      addressClass,
      algorithm,
      payloadHex: payload.toString('hex'),
    }),
  };
}

export function addressesEqual(left: BlockchainAddress, right: BlockchainAddress): boolean {
  return left.binaryHex === right.binaryHex;
}

File: packages/sunrey-chain/rust/crates/wallet/src/address.rs
//! Versioned SunRey Blockchain address format.
//!
//! Canonical binary is 42 bytes. Text is HRP + '1' + RFC 4648 base32.
//! Network class, algorithm, and a SHA-256 descriptor payload are bound
//! into the address. Cross-network reuse is rejected.

use sha2::{Digest, Sha256};

pub const ADDRESS_VERSION: u8 = 1;
pub const ADDRESS_MAX_BINARY: usize = 42;
pub const ADDRESS_MAX_TEXT: usize = 90;
const MAGIC: &[u8; 2] = b"SR";
const PAYLOAD_DOMAIN: &[u8] = b"SUNREY-ADDR-PAYLOAD-V1";
const CHECKSUM_DOMAIN: &[u8] = b"SUNREY-ADDR-V1";
const BASE32: &[u8] = b"abcdefghijklmnopqrstuvwxyz234567";

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NetworkClass {
    Development = 1,
    ReservedTest = 2,
    ReservedProduction = 3,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AddressClass {
    SingleKey = 1,
    Policy = 2,
    MultiAuth = 3,
    Machine = 4,
    Institutional = 5,
    WatchOnly = 6,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AddressAlgorithm {
    Ed25519V1 = 1,
    HybridSimV1 = 2,
    PqSimV1 = 3,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BlockchainAddress {
    pub text: String,
    pub binary: Vec<u8>,
    pub network_id: String,
    pub network_class: NetworkClass,
    pub address_class: AddressClass,
    pub algorithm: AddressAlgorithm,
    pub payload: [u8; 32],
}

#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum AddressError {
    #[error("wrong-network address")]
    WrongNetwork,
    #[error("address checksum failed")]
    ChecksumFailure,
    #[error("malformed address")]
    Malformed,
    #[error("unknown address version")]
    UnknownVersion,
    #[error("unknown address class")]
    UnknownClass,
    #[error("unknown address algorithm")]
    UnknownAlgorithm,
}

impl NetworkClass {
    pub fn from_network_id(network_id: &str) -> Option<Self> {
        if matches!(network_id, "net_sunrey_simulation" | "net_sunrey_local_dev") {
            return Some(Self::Development);
        }
        if network_id == "net_sunrey_reserved_test" || network_id.starts_with("net_sunrey_testnet_")
        {
            return Some(Self::ReservedTest);
        }
        if network_id == "net_sunrey_reserved_production" {
            return Some(Self::ReservedProduction);
        }
        None
    }

    pub fn hrp(self) -> &'static str {
        match self {
            Self::Development => "srdev",
            Self::ReservedTest => "srtst",
            Self::ReservedProduction => "srprd",
        }
    }

    pub fn from_hrp(hrp: &str) -> Option<Self> {
        match hrp {
            "srdev" => Some(Self::Development),
            "srtst" => Some(Self::ReservedTest),
            "srprd" => Some(Self::ReservedProduction),
            _ => None,
        }
    }

    pub fn from_u8(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::Development),
            2 => Some(Self::ReservedTest),
            3 => Some(Self::ReservedProduction),
            _ => None,
        }
    }
}

impl AddressClass {
    pub fn from_u8(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::SingleKey),
            2 => Some(Self::Policy),
            3 => Some(Self::MultiAuth),
            4 => Some(Self::Machine),
            5 => Some(Self::Institutional),
            6 => Some(Self::WatchOnly),
            _ => None,
        }
    }
}

impl AddressAlgorithm {
    pub fn from_u8(value: u8) -> Option<Self> {
        match value {
            1 => Some(Self::Ed25519V1),
            2 => Some(Self::HybridSimV1),
            3 => Some(Self::PqSimV1),
            _ => None,
        }
    }
}

fn sha256(parts: &[&[u8]]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    for part in parts {
        hasher.update(part);
    }
    hasher.finalize().into()
}

pub fn descriptor_payload(
    network_id: &str,
    class: AddressClass,
    algorithm: AddressAlgorithm,
    descriptor: &[u8],
) -> [u8; 32] {
    sha256(&[PAYLOAD_DOMAIN, network_id.as_bytes(), &[class as u8], &[algorithm as u8], descriptor])
}

fn encode_base32(bytes: &[u8]) -> String {
    let mut bits = 0u32;
    let mut value = 0u32;
    let mut out = String::new();
    for byte in bytes {
        value = (value << 8) | u32::from(*byte);
        bits += 8;
        while bits >= 5 {
            let idx = ((value >> (bits - 5)) & 31) as usize;
            out.push(BASE32[idx] as char);
            bits -= 5;
        }
    }
    if bits > 0 {
        let idx = ((value << (5 - bits)) & 31) as usize;
        out.push(BASE32[idx] as char);
    }
    out
}

fn decode_base32(text: &str) -> Option<Vec<u8>> {
    let mut bits = 0u32;
    let mut value = 0u32;
    let mut out = Vec::new();
    for ch in text.bytes() {
        let idx = BASE32.iter().position(|item| *item == ch)?;
        value = (value << 5) | idx as u32;
        bits += 5;
        if bits >= 8 {
            out.push(((value >> (bits - 8)) & 255) as u8);
            bits -= 8;
        }
    }
    Some(out)
}

fn checksum(prefix: &[u8]) -> [u8; 4] {
    let digest = sha256(&[CHECKSUM_DOMAIN, prefix]);
    [digest[0], digest[1], digest[2], digest[3]]
}

pub fn encode_address(
    network_id: &str,
    class: AddressClass,
    algorithm: AddressAlgorithm,
    descriptor: &[u8],
) -> Result<BlockchainAddress, AddressError> {
    let network_class =
        NetworkClass::from_network_id(network_id).ok_or(AddressError::WrongNetwork)?;
    let payload = descriptor_payload(network_id, class, algorithm, descriptor);
    let mut prefix = [0u8; 38];
    prefix[0..2].copy_from_slice(MAGIC);
    prefix[2] = ADDRESS_VERSION;
    prefix[3] = network_class as u8;
    prefix[4] = class as u8;
    prefix[5] = algorithm as u8;
    prefix[6..38].copy_from_slice(&payload);
    let sum = checksum(&prefix);
    let mut binary = prefix.to_vec();
    binary.extend_from_slice(&sum);
    if binary.len() != ADDRESS_MAX_BINARY {
        return Err(AddressError::Malformed);
    }
    let mut body = vec![ADDRESS_VERSION, class as u8, algorithm as u8];
    body.extend_from_slice(&payload);
    body.extend_from_slice(&sum);
    let text = format!("{}1{}", network_class.hrp(), encode_base32(&body));
    if text.len() > ADDRESS_MAX_TEXT {
        return Err(AddressError::Malformed);
    }
    Ok(BlockchainAddress {
        text,
        binary,
        network_id: network_id.to_string(),
        network_class,
        address_class: class,
        algorithm,
        payload,
    })
}

pub fn parse_address(
    text: &str,
    expected_network_id: Option<&str>,
) -> Result<BlockchainAddress, AddressError> {
    if text.len() > ADDRESS_MAX_TEXT {
        return Err(AddressError::Malformed);
    }
    let sep = text.find('1').ok_or(AddressError::Malformed)?;
    let hrp = &text[..sep];
    let body = &text[sep + 1..];
    let network_class = NetworkClass::from_hrp(hrp).ok_or(AddressError::WrongNetwork)?;
    let decoded = decode_base32(body).ok_or(AddressError::Malformed)?;
    if decoded.len() < 39 {
        return Err(AddressError::Malformed);
    }
    if decoded[0] != ADDRESS_VERSION {
        return Err(AddressError::UnknownVersion);
    }
    let class = AddressClass::from_u8(decoded[1]).ok_or(AddressError::UnknownClass)?;
    let algorithm = AddressAlgorithm::from_u8(decoded[2]).ok_or(AddressError::UnknownAlgorithm)?;
    let mut payload = [0u8; 32];
    payload.copy_from_slice(&decoded[3..35]);
    let given = &decoded[35..39];
    let mut prefix = [0u8; 38];
    prefix[0..2].copy_from_slice(MAGIC);
    prefix[2] = ADDRESS_VERSION;
    prefix[3] = network_class as u8;
    prefix[4] = class as u8;
    prefix[5] = algorithm as u8;
    prefix[6..38].copy_from_slice(&payload);
    if given != checksum(&prefix) {
        return Err(AddressError::ChecksumFailure);
    }
    if let Some(expected) = expected_network_id {
        let expected_class =
            NetworkClass::from_network_id(expected).ok_or(AddressError::WrongNetwork)?;
        if expected_class != network_class {
            return Err(AddressError::WrongNetwork);
        }
    }
    let mut binary = prefix.to_vec();
    binary.extend_from_slice(given);
    Ok(BlockchainAddress {
        text: text.to_string(),
        binary,
        network_id: expected_network_id.unwrap_or("net_sunrey_simulation").to_string(),
        network_class,
        address_class: class,
        algorithm,
        payload,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_and_checksum() {
        let addr = encode_address(
            "net_sunrey_simulation",
            AddressClass::SingleKey,
            AddressAlgorithm::Ed25519V1,
            b"alice-public",
        )
        .unwrap();
        assert!(addr.text.starts_with("srdev1"));
        let parsed = parse_address(&addr.text, Some("net_sunrey_simulation")).unwrap();
        assert_eq!(parsed.payload, addr.payload);
        let mut broken = addr.text.clone();
        broken.replace_range(broken.len() - 1.., if broken.ends_with('a') { "b" } else { "a" });
        assert_eq!(
            parse_address(&broken, Some("net_sunrey_simulation")),
            Err(AddressError::ChecksumFailure)
        );
    }

    #[test]
    fn testnet_hrp_is_srtst() {
        let addr = encode_address(
            "net_sunrey_testnet_1",
            AddressClass::SingleKey,
            AddressAlgorithm::Ed25519V1,
            b"alice-public",
        )
        .unwrap();
        assert!(addr.text.starts_with("srtst1"));
        let parsed = parse_address(&addr.text, Some("net_sunrey_testnet_1")).unwrap();
        assert_eq!(parsed.network_class, NetworkClass::ReservedTest);
    }

    #[test]
    fn wrong_network_rejected() {
        let addr = encode_address(
            "net_sunrey_reserved_production",
            AddressClass::SingleKey,
            AddressAlgorithm::Ed25519V1,
            b"alice-public",
        )
        .unwrap();
        assert_eq!(
            parse_address(&addr.text, Some("net_sunrey_simulation")),
            Err(AddressError::WrongNetwork)
        );
    }
}
